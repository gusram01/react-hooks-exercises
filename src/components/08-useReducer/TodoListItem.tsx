import React from 'react';
import { TodoState } from './TodoReducer';
import PropTypes from 'prop-types';

interface Props {
  todo: Partial<TodoState>;
  i: number;
  handleToggle: (id: number) => void;
  handleDelete: (id: number) => void;
}

const TodoListItem: React.FC<Props> = ({ todo, i, handleToggle, handleDelete }) => {

  return (
    <li
      key={todo.id}
      className="list-group-item"
    >
      <p
        className={`${todo.done && 'complete'}`}
        onClick={() => handleToggle(todo.id!)}
      >
        {i + 1}.- {todo.desc}
      </p>
      <button
        className="btn btn-danger"
        onClick={() => handleDelete(todo.id!)}
      >
        Delete todo
      </button>
    </li>
  );
};

TodoListItem.propTypes = {
  todo: PropTypes.object.isRequired,
  i: PropTypes.number.isRequired,
  handleToggle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired
}

export default TodoListItem;
