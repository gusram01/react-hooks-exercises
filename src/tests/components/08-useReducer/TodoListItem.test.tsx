import React from 'react';
import TodoListItem from '../../../components/08-useReducer/TodoListItem';
import { shallow, ShallowWrapper } from 'enzyme';
import { dummyTodo } from '../../fixtures/dummyTodos';


describe('TodoListItem tests', () => {
  const handleToggle = jest.fn();
  const handleDelete = jest.fn();
  const wrapper = shallow(
    <TodoListItem
      todo={dummyTodo[0]}
      i={1}
      handleToggle={handleToggle}
      handleDelete={handleDelete}
    />
  );

  test('TodoListItem should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('handleDelete should been called', () => {
    wrapper.find('.btn-danger').simulate('click');
    expect(handleDelete).toHaveBeenCalledWith(1);
  });

  test('handleToggle should been called', () => {
    wrapper.find('p').simulate('click');
    expect(handleToggle).toHaveBeenCalledWith(1);
  });

  test('TodoListItem should render the todo inner  text correctly', () => {
    expect(wrapper.find('p').text()).toBe('2.- Learn React');
  });

  test('Should change class style if done is checked', () => {
    const todo = dummyTodo[0];
    todo.done = true;
    const wrapper = shallow(
      <TodoListItem
        todo={todo}
        i={1}
        handleToggle={handleToggle}
        handleDelete={handleDelete}
      />
    );
    expect(wrapper.find('p').hasClass('complete')).toBe(true);
  });

});