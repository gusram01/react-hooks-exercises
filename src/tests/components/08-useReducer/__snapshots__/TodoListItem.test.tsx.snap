// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem tests TodoListItem should render correctly 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false"
    onClick={[Function]}
  >
    2
    .- 
    Learn React
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete todo
  </button>
</li>
`;
